### DISCUSSION TITLE
[Native Hosting] Can A make two-way calls?

### DISCUSSION THREAD

**Discussion by ChocolaDiGiCharat** [dotnet/runtime#78799] [Q&A] [answered]
I have learned how to call a .net assembly from a C++ app on Linux using Native Hosting.

I can call methods in the assembly from C++, but is there a way to call methods exposed by the HOST (C++) from the assembly?

Or is there an alternative method for that?

**Comment by hez2010** [comment, marked as answer]
You can use function pointer to call C++ code from C# directly:

C# side:

```csharp
unsafe void Foo(IntPtr ptr)
{
    var f = (delegate* unmanaged[Cdecl]<int, int, int>)ptr;
    Console.WriteLine(f(4, 5));
}
```

C++ side:
```cpp
int f(int x, int y) { return x + y; }

lib_args args { &f };
Foo(&args, sizeof(args));
```

**Reply by ChocolaDiGiCharat** [reply]
Thank you very much! It worked fine.
